File: book/snippets/nostr/js/src/nip19.js
const { loadWasmSync, Keys, EventBuilder, Nip19Profile, Nip19Event, Nip19Relay} = require("@rust-nostr/nostr");

function run() {
    // Load WASM
    loadWasmSync();

    // Generate random keys
    let keys = Keys.generate();

    console.log();
    console.log("Bare keys and ids (bech32):");
    // ANCHOR: nip19-npub
    console.log(` Public key: ${keys.publicKey.toBech32()}`);
    // ANCHOR_END: nip19-npub

    // ANCHOR: nip19-nsec
    console.log(` Secret key: ${keys.secretKey.toBech32()}`);
    // ANCHOR_END: nip19-nsec

    // ANCHOR: nip19-note
    let event = EventBuilder.textNote("Hello from Rust Nostr JS Bindings!", []).toEvent(keys);
    console.log(` Event     : ${event.id.toBech32()}`);
    // ANCHOR_END: nip19-note

    console.log();
    console.log("Shareable identifiers with extra metadata (bech32):");
    // ANCHOR: nip19-nprofile-encode
    // Create NIP-19 profile including relays data
    let relays = ["wss://relay.damus.io"];
    let nprofile = new Nip19Profile(keys.publicKey, relays);
    console.log(` Profile (encoded): ${nprofile.toBech32()}`);
    // ANCHOR_END: nip19-nprofile-encode

    // ANCHOR: nip19-nprofile-decode
    // Decode NIP-19 profile
    let decode_nprofile = Nip19Profile.fromBech32(nprofile.toBech32());
    console.log(` Profile (decoded): ${decode_nprofile.publicKey().toBech32()}`);
    // ANCHOR_END: nip19-nprofile-decode

    console.log();
    // ANCHOR: nip19-nevent-encode
    // Create NIP-19 event including author and relays data
    let nevent = new Nip19Event(event.id, keys.publicKey, null, relays);
    console.log(` Event (encoded): ${nevent.toBech32()}`);
    // ANCHOR_END: nip19-nevent-encode

    // ANCHOR: nip19-nevent-decode
    // Decode NIP-19 event
    let decode_nevent = Nip19Event.fromBech32(nevent.toBech32());
    console.log(` Event (decoded): ${decode_nevent.eventId().toBech32()}`);
    // ANCHOR_END: nip19-nevent-decode

    console.log();
    // ANCHOR: nip19-nrelay-encode
    // Create NIP-19 relay
    let relay = new Nip19Relay("wss://relay.damus.io");
    console.log(` Relay (encoded): ${relay.toBech32()}`);
    // ANCHOR_END: nip19-nrelay-encode

    // ANCHOR: nip19-nrelay-decode
    // Decode NIP-19 relay
    let decode_relay = Nip19Relay.fromBech32(relay.toBech32());
    console.log(` Relay (decoded): ${decode_relay.url()}`);
    // ANCHOR_END: nip19-nrelay-decode

}

module.exports.run = run;

File: book/snippets/python/src/nip19.py
from nostr_sdk import Keys, EventBuilder, Nip19Profile, Nip19, Nip19Event, Coordinate, Kind
def nip19():
    keys = Keys.generate()

    print()
    print("Bare keys and ids (bech32):")
    # ANCHOR: nip19-npub
    print(f" Public key: {keys.public_key().to_bech32()}")
    # ANCHOR_END: nip19-npub

    # ANCHOR: nip19-nsec
    print(f" Secret key: {keys.secret_key().to_bech32()}")
    # ANCHOR_END: nip19-nsec

    # ANCHOR: nip19-note
    event = EventBuilder.text_note("Hello from rust-nostr Python bindings!").sign_with_keys(keys)
    print(f" Event     : {event.id().to_bech32()}")
    # ANCHOR_END: nip19-note

    print()
    print("Shareable identifiers with extra metadata (bech32):")
    # ANCHOR: nip19-nprofile-encode
    # Create NIP-19 profile including relays data
    relays = ["wss://relay.damus.io"]
    nprofile = Nip19Profile(keys.public_key(),relays)
    print(f" Profile (encoded): {nprofile.to_bech32()}")
    # ANCHOR_END: nip19-nprofile-encode

    # ANCHOR: nip19-nprofile-decode
    # Decode NIP-19 profile
    decode_nprofile = Nip19.from_bech32(nprofile.to_bech32())
    print(f" Profile (decoded): {decode_nprofile}")
    # ANCHOR_END: nip19-nprofile-decode

    print()
    # ANCHOR: nip19-nevent-encode
    # Create NIP-19 event including author and relays data
    nevent = Nip19Event(event.id(), keys.public_key(), kind=None, relays=relays)
    print(f" Event (encoded): {nevent.to_bech32()}")
    # ANCHOR_END: nip19-nevent-encode

    # ANCHOR: nip19-nevent-decode
    # Decode NIP-19 event
    decode_nevent = Nip19.from_bech32(nevent.to_bech32())
    print(f" Event (decoded): {decode_nevent}")
    # ANCHOR_END: nip19-nevent-decode

    print()
    # ANCHOR: nip19-naddr-encode
    # Create NIP-19 coordinate
    coord = Coordinate(Kind(0),keys.public_key())
    print(f" Coordinate (encoded): {coord.to_bech32()}")
    # ANCHOR_END: nip19-naddr-encode

    # ANCHOR: nip19-naddr-decode
    # Decode NIP-19 coordinate
    decode_coord = Nip19.from_bech32(coord.to_bech32())
    print(f" Coordinate (decoded): {decode_coord}")
    # ANCHOR_END: nip19-naddr-decode
